pallets/stateful-storage/src/stateful_signed_extension.rs

```diff
@@ -0,0 +1,284 @@
+use crate::{types::ItemAction, Call, Config, Pallet};
+use core::marker::PhantomData;
+
+use codec::{Decode, Encode};
+use common_primitives::{
+	msa::MessageSourceId,
+	schema::{PayloadLocation, SchemaId},
+	stateful_storage::{PageHash, PageId},
+};
+use frame_support::{
+	dispatch::{DispatchInfo, Dispatchable},
+	pallet_prelude::ValidTransaction,
+	traits::IsSubType,
+	BoundedVec,
+};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{DispatchInfoOf, SignedExtension},
+	transaction_validity::{InvalidTransaction, TransactionValidity, TransactionValidityError},
+	DispatchError,
+};
+
+/// The SignedExtension trait is implemented on CheckFreeExtrinsicUse to validate that a provider
+/// has not already been revoked if the calling extrinsic is revoking a provider to an MSA. The
+/// purpose of this is to ensure that the revoke_delegation_by_delegator extrinsic cannot be
+/// repeatedly called and flood the network.
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct StatefulSignedExtension<T: Config + Send + Sync>(PhantomData<T>);
+
+pub trait StatefulStorageValidate<C: Config> {
+	fn validate_itemized(
+		account_id: &C::AccountId,
+		msa_id: &MessageSourceId,
+		schema_id: &SchemaId,
+		actions: &BoundedVec<ItemAction, C::MaxItemizedActionsCount>,
+		is_payload_signed: bool,
+		target_hash: &PageHash,
+	) -> TransactionValidity;
+
+	fn validate_paginated(
+		account_id: &C::AccountId,
+		msa_id: &MessageSourceId,
+		schema_id: &SchemaId,
+		page_id: &PageId,
+		is_payload_signed: bool,
+		page_delete: bool,
+		target_hash: &PageHash,
+	) -> TransactionValidity;
+}
+
+impl<T: Config> StatefulStorageValidate<T> for Pallet<T> {
+	/// Validates the following criteria for an Itemized Stateful Storage extrinsic:
+	///
+	/// * schema_id represents a valid schema
+	/// * schema `PayloadLocation` is `Paginated`
+	/// * schema allows signed payloads if `is_payload_signed` == true
+	/// * schema is not `AppendOnly` if `page_delete` == true
+	/// * page can be read/parsed
+	/// * known page state hash matches provided target hash
+	///
+	/// Returns a `ValidTransaction` or wrapped [`pallet::Error`]
+	///
+	/// # Arguments:
+	/// * account_id: the account id associated with the MSA owning the storage
+	/// * msa_id: the `MessageSourceAccount`
+	/// * schema_id: the `SchemaId` for the Paginated storage
+	/// * actions: Array of `ItemAction` representing adds/deletes to Itemized storage
+	/// * is_payload_signed: whether we were called with a signed payload
+	/// * target_hash: the content hash representing the known state of the Paginated storage.
+	///
+	/// # Errors (as u8 wrapped by `InvalidTransaction::Custom`)
+	/// * [`Error::InvalidSchemaId`]
+	/// * [`Error::SchemaPayloadLocationMismatch`]
+	/// * [`Error::SchemaNotSupported`]
+	/// * [`Error::CorruptedState`]
+	/// * [`Error::StalePageState`]
+	///
+	fn validate_itemized(
```

Comment:
Now that I'm thinking about itemized this would generate 2 READ in weights for reading itemized and that's not so great. This works for paginated since we don't need to read the item for modification but for itemized it is problematic.

Reply:
maybe we just check the hash for itemized in extrinsic? It would be different from paginated but not sure how else we can keep the weights optimized.

Reply:
As far as I can tell, the number of reads hasn't changed; only the order they're done in. Yes, we read Itemized twice, but the second read is from Overlay cache.

Reply:
It is but weights don't know about overlay so it would appear as 2 reads in weights and block space would be wasted.

Reply:
Here are the reads the benchmark sees for `Itemized`:
```
	// Storage: Schemas Schemas (r:1 w:0)
	// Storage: Msa PublicKeyToMsaId (r:1 w:0)
	// Storage: Msa DelegatorAndProviderToDelegation (r:1 w:0)
	// Storage: unknown [0xbd1557c8db6bd8599a811a7175fbc2fc6400] (r:1 w:1)
```

Reply:
Hmm, it is weird that it just showed up as 1 read instead of 2 reads. Maybe they fixed that overlay part for weights? I'm a bit skeptical 🤔

Reply:
I hope there is no mock call that is hiding one read but otherwise looks like cache works